### PR TITLE
Add: pagination functionality

### PR DESCRIPTION
**Pull request by ankkit24**

### CODE REVIEW

File: src/main/java/com/redhat/pantheon/servlet/ModuleListingServlet.java [hunk 1]
```diff
@@ -39,16 +39,27 @@
     protected void doGet(@Nonnull SlingHttpServletRequest request, @Nonnull SlingHttpServletResponse response) throws ServletException, IOException {
         ModuleDataRetriever mdr = new ModuleDataRetriever(request.getResourceResolver());
         String searchParam = getParam(request, "search");
+        String countParam = getParam(request, "search");
```
Comment (benradey) [inline src/main/java/com/redhat/pantheon/servlet/ModuleListingServlet.java]: I suspect this is a typo and that you intended to get the param of `count` instead, but I think we shouldn't need a count param at all. I see that this variable is unused, so please remove it.

File: src/main/java/com/redhat/pantheon/servlet/ModuleListingServlet.java [hunk 2]
```diff
@@ -39,16 +39,27 @@
...
         String directionParam = getParam(request, "direction");
         String offset = getParam(request, "offset");
         String limit = getParam(request, "limit");
 
         try {
-            List<Map<String, Object>> payload = mdr.getModulesSort(searchParam, keyParam, directionParam, offset, limit);
             response.setContentType("application/json");
             Writer w = response.getWriter();
-            w.write(new ObjectMapper().writer().withDefaultPrettyPrinter().writeValueAsString(payload));
+            if(searchParam != ""){
```
Comment (benradey) [inline src/main/java/com/redhat/pantheon/servlet/ModuleListingServlet.java]: I'm not quite sure what's happening with this `if` statement, but it doesn't seem right. We shouldn't be only returning the row count depending on whether the user supplied a search parameter. Let's talk on Monday about how this ought to look.

File: src/main/java/com/redhat/pantheon/data/ModuleDataRetriever.java [hunk 3]
```diff
@@ -59,6 +59,20 @@ public ModuleDataRetriever(ResourceResolver resolver) {
...
+    public int getRowCount() throws Exception{
+            // String query = "AND (a.[jcr:title] like " + "'%" + query + "%' " +
+            //         "OR a.[jcr:description] like " + "'%" + query + "%') ";
+
+            StringBuilder countQuery = new StringBuilder()
+                .append("select * from [nt:base] as a ")
+                .append("where [sling:resourceType] = 'pantheon/modules' ")
+                .append("and (isdescendantnode(a, '/content/repositories') ")
+                .append("or isdescendantnode(a, '/content/modules') ")
+                .append("or isdescendantnode(a, '/content/sandboxes')) ");
```
[the remark this answers is not in the document]
Comment (carlosmunoz) [reply src/main/java/com/redhat/pantheon/data/ModuleDataRetriever.java]: Just a quick comment on this: I agree that `queryRaw` is not the right name for this method, and that we need the where clause in the count query.
As for the `select count()` part, unfortunately there is no way to do that in JCR-QL. This approach (calling `getRowCount()`) is the only way we have to get the total count. I *think* getting the row count does not imply fetching everything in memory. If I'm mistaken, we might need to think of a smart way to do the pagination.

**Comment by benradey** [comment]
Holy smokes, you're right!

https://stackoverflow.com/questions/26400206/how-to-apply-aggregate-functionslike-min-max-count-in-jcr-sql2

Looks like the only way to get the count is manually in the java code. That
stinks, but if it's what we have to do, then so be it...

File: src/main/java/com/redhat/pantheon/data/ModuleDataRetriever.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (benradey) [reply src/main/java/com/redhat/pantheon/data/ModuleDataRetriever.java]: Holy smokes, you're right!

https://stackoverflow.com/questions/26400206/how-to-apply-aggregate-functionslike-min-max-count-in-jcr-sql2

Looks like the only way to get the count is manually in the java code. That stinks, but if it's what we have to do, then so be it...

File: src/main/java/com/redhat/pantheon/jcr/JcrQueryHelper.java [hunk 4]
```diff
@@ -51,6 +53,23 @@ public JcrQueryHelper(ResourceResolver resourceResolver) {
...
+        Query queryObj = queryManager.createQuery(query, Query.JCR_SQL2);
+        QueryResult result = queryObj.execute();
+        NodeIterator nodes = result.getNodes();
+        
+        int rowCount = 0;
+        while (nodes.hasNext()) {
+            Node node = nodes.nextNode();
+            rowCount+=1;
+        }
+        return rowCount;
```
[the remark this answers is not in the document]
Comment (benradey) [reply src/main/java/com/redhat/pantheon/jcr/JcrQueryHelper.java]: Oops! There is no `select count()`, nevermind... :(

**Comment by jmarrero** [comment]
@ankkit24 this needs to be rebased on top of upstream/master.

File: frontend/src/app/Pagination.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import React, { Fragment } from 'react';
+import {Button} from '@patternfly/react-core';
+import { Badge } from '@patternfly/react-core';
+
+export interface IProps{
```
Comment (jmarrero) [inline frontend/src/app/Pagination.tsx]: Is  "newPage: number" just the page number? If so I would rename that to pageNumber instead 

and "isNextPageRequied: number" is named like a boolean but it's expecting a number. I would either turn this into a boolean or rename it to represent what it is really doing.

File: frontend/src/app/search.tsx [hunk 6]
```diff
@@ -178,41 +203,42 @@ export default class Search extends Component {
...
     console.log('handleSelectAll')
     this.setState({check: !this.state.check}, () => {
       this.setState(prevState => {
+        this.state.allPaths=[]
+        this.transientPaths=[]
         const selectAllcheck = this.state.data.map(dataitem => {
               dataitem["checkedItem"] = this.state.check
               console.log(dataitem["pant:transientPath"]+":"+dataitem["checkedItem"])
-              this.state.check?this.state.allPaths.push(dataitem["pant:transientPath"]):delete this.state.allPaths[this.state.allPaths.indexOf(dataitem["pant:transientPath"])]
+              this.state.check?this.state.allPaths.push(dataitem["pant:transientPath"]):this.state.allPaths.splice(this.state.allPaths.indexOf(dataitem["pant:transientPath"]),1)
```
Comment (jmarrero) [inline frontend/src/app/search.tsx]: I see some of the checkbox changes here, make sure you don't overwrite your checkboxes changes that are already on Master. When doing the rebase the new change set should only show the stuff for the pagination.

**Comment by jmarrero** [comment]
It builds
basic functionality works
functions added verified
I can follow the code
test coverage went up

File: frontend/src/app/Pagination.tsx (hunk 5 above)
Comment (ankkit24) [reply frontend/src/app/Pagination.tsx]: Hey Joseph!
I have made the change of name for this variable to 'nextPageRecordCount' which indicates the number of records in the next page.

**Pull request merged**